### PR TITLE
SI-6331 deconst If type / refine equality of floating point Constant types.

### PR DESCRIPTION
**Pull request by retronym**
Review by @paulp

/cc @xeno-by 
1. I'd suggest you remove `ConstantType#isNaN` from `scala.reflect.api.Constants#ConstantApi`.
2. Do you expect this output when compiling a macro in partest?

```
quick.partest:
[scalacfork] Compiling 1 file to /Users/jason/code/scala/build/quick/classes/partest
[scalacfork] // produced from source-/Users/jason/code/scala/src/partest/scala/tools/partest/package.scala,line-98,offset=2789
[scalacfork] object Test extends App {
[scalacfork]   val $u: c.universe.type = c.universe;
[scalacfork]   import $u._
[scalacfork]   val tpe = evidence$1.in($m).tpe
[scalacfork]   println(tpe)
[scalacfork] }
[scalacfork] // produced from source-/Users/jason/code/scala/src/partest/scala/tools/partest/package.scala,line-98,offset=2789
[scalacfork] object Test extends App {
[scalacfork]   val $u: c.universe.type = c.universe;
[scalacfork]   val $m: $u.Mirror = c.universe.rootMirror;
[scalacfork]   import $u._, $m._
[scalacfork]   val tree = Block(List(Apply(Select(Select(build.This(staticPackage("scala").asModule.moduleClass), newTermName("Predef")), newTermName("println")), List(Apply(Select(Apply(Select(Select(build.Ident(staticPackage("scala")), newTermName("StringContext")), newTermName("apply")), List(Literal(Constant("trace> ")), Literal(Constant("\\nres: ")), Literal(Constant(" = ")), Literal(Constant("\\n")))), newTermName("s")), List(exprCode.in($m).tree, exprType.in($m).tree, a.in($m).tree))))), a.in($m).tree)
[scalacfork]   println(tree)
[scalacfork] }
[propertyfile] Updating property file: /Users/jason/code/scala/build/quick/classes/partest/partest.properties
[stopwatch] [quick.partest.timer: 10.808 sec]
```

### CODE REVIEW

File: src/partest/scala/tools/partest/package.scala [hunk 1]
```diff
@@ -73,4 +73,31 @@ package object partest {
```
Comment (paulp) [inline src/partest/scala/tools/partest/package.scala:72]: I'm thinking scala/tools/nsc/package.scala.  Why punish ourselves by hiding it in partest.

Comment (retronym) [inline src/partest/scala/tools/partest/package.scala:72]: That might take an expert hand to wire it into the fast track mechanism. Let's do that separately from this PR.

**Comment by retronym** [comment]
PLS REBUILD ALL  :pray:

**Comment by retronym** [comment]
In order to clean up `Constant#hashCode`, I need to understand `Constant#tag`. It was added to `Constant#equals` way back when in 4720d45a83 without adding it to `hashCode` (EDIT: which doesn't break the contract); `Constant#tag` itself was added even longer ago in the dark days of CVS 4ae459ef.

My question: Is there an example where `this.value == other.value && this.tag != other.tag`? If not, why is `tag` considered in `equals`? Penny for your thoughts, @odersky?

File: src/partest/scala/tools/partest/package.scala (hunk 1 above)
Comment (xeno-by) [inline src/partest/scala/tools/partest/package.scala:72]: There's a very good example of integrating a "normal" macro into fast track:
1) macro (you already did this step): https://github.com/scala/scala/blob/2.10.x/src/reflect/scala/reflect/runtime/package.scala
2) registration in definitions: https://github.com/scala/scala/blob/2.10.x/src/reflect/scala/reflect/internal/Definitions.scala#L489
3) bridge in fast track: https://github.com/scala/scala/blob/2.10.x/src/compiler/scala/tools/reflect/FastTrack.scala#L37

Nothing scary. Just two extra lines and a redeployment of starr.

**Comment by retronym** [comment]
PLS REBUILD ALL

**Comment by gkossakowski** [comment]
What's the status of this one? Shall commit breaking build be squashed with ones fixing it? Do we need to wait for @odersky's feedback?

**Comment by retronym** [comment]
All the build failures above were spurious (the build didn't even start).

All commits in the sequence actually build. I forgot to amend `hashCode`, but it still worked because `(-0d).## == 0d.##`. I added a test to show this, then went and made `hashCode` more obviously correct, and also better distributed to boot.

In short: I don't think we need to squash, and I'm happy with all the commits. The open question I had ("why does equals bother to compare `tag`) is noted in a comment.

**Comment by jsuereth** [comment]
Yeah, I think this one is good to go as is, just with follow on work.

**Comment by jsuereth** [comment]
@paulp or @xeno-by  can I get a LGTM here?

**Comment by paulp** [comment]
What kind of nutbag would find something with this kind of test coverage to look anything less than good.  LGTM.

You sound like you're ready to lead some religious revivals.

```
"Can I get an LGTM?"
"LGTM!"
"Can I get a hallelujah, LGTM!"
"Hallelujah, LGTM!"
"Can I get a test case?"
"Not as a general rule, no."
```

**Pull request merged**